### PR TITLE
Delete domain IP addresses from networks

### PR DESCRIPTION
**Pull request by eamonnotoole**
We've found that, when executing destroy/create cycles on
a domain we get network failures since the old entries for
the domain are still present in the virtual network(s). We've
also found that, on occasion, we don't get complete information
on a domain's interfaces from ListAllInterfaceAddresses. This
patch implements a couple of changes in light of these issues:

1.  We now remove the dhcp entries for a domain when the domain
    is deleted
2.  We check the number of networks returned from getDomainInterfaces
    against the number of networks that we see in the domain XML.
    If the second number is greater than the first then we get
    interface information directly from the networks attached to
    the domain.

### CODE REVIEW

**Comment by inercia** [comment]
>  when executing destroy/create cycles on a domain we get network failures since the old entries for the domain are still present in the virtual network(s). 

@eamonnotoole So you mean that, even when a network is deleted, the DHCP server is still running? Or you mean that some configuration is remaining somewhere, and when a new network is created it takes that configuration somehow? Are you sure the network has been destroyed?

File: libvirt/resource_libvirt_domain.go [hunk 1]
```diff
@@ -996,3 +1048,56 @@ func getDomainInterfaces(domain *libvirt.VirDomain) ([]libvirt.VirDomainInterfac
 
 	return interfaces, nil
 }
+
+func getDomainInterfacesFromNetworks(domain defDomain, meta interface{}) ([]libvirt.VirDomainInterface, error) {
+
+	log.Printf("[DEBUG] getting domain addresses from networks")
+	var virDomain = make([]libvirt.VirDomainInterface, 0)
```
Comment (inercia) [inline libvirt/resource_libvirt_domain.go]: `virDomain` looks like a _domain_, but it is really a _list of interfaces_. What do you think about `ifacesLst := make([]libvirt.VirDomainInterface, 0)`?

Comment (inercia) [inline libvirt/resource_libvirt_domain.go]: In fact, I think you wouldn't need the `make` at all as you are `append`ing, right?

File: libvirt/resource_libvirt_domain.go [hunk 2]
```diff
@@ -996,3 +1048,56 @@ func getDomainInterfaces(domain *libvirt.VirDomain) ([]libvirt.VirDomainInterfac
 
 	return interfaces, nil
 }
+
+func getDomainInterfacesFromNetworks(domain defDomain, meta interface{}) ([]libvirt.VirDomainInterface, error) {
```
Comment (inercia) [inline libvirt/resource_libvirt_domain.go]: This function is really similar to [`getDomainInterfacesViaQemuAgent`](https://github.com/dmacvicar/terraform-provider-libvirt/blob/master/libvirt/qemu_agent.go#L32). It is a pity they do not have the same function signature...

Shouldn't we invoke it from [inside `getDomainInterfaces`](https://github.com/dmacvicar/terraform-provider-libvirt/blob/master/libvirt/resource_libvirt_domain.go#L979)? So we could try to get the interfaces with the qemu agent, then use `getDomainInterfacesFromNetworks` if it fails, and then try with `ListAllInterfaceAddresses` as a last resource...

File: libvirt/resource_libvirt_domain.go [hunk 3]
```diff
@@ -699,6 +699,25 @@ func resourceLibvirtDomainRead(d *schema.ResourceData, meta interface{}) error {
...
+		interfacesIfaces[virDom.Name] = true
+	}
+
+	if len(interfacesXML) > len(interfacesIfaces) {
+		ifacesWithAddr, err = getDomainInterfacesFromNetworks(domainDef, meta)
+		if err != nil {
+			return fmt.Errorf("Error retrieving interface addresses: %s", err)
+		}
+	}
+
```
Comment (inercia) [inline libvirt/resource_libvirt_domain.go]: Maybe we could move this whole block [inside `getDomainInterfaces`](https://github.com/dmacvicar/terraform-provider-libvirt/blob/master/libvirt/resource_libvirt_domain.go#L979)...

File: libvirt/resource_libvirt_domain.go [hunk 4]
```diff
@@ -858,6 +877,39 @@ func resourceLibvirtDomainDelete(d *schema.ResourceData, meta interface{}) error
...
+		addresses := d.Get(prefix + ".addresses")
+		for _, addressI := range addresses.([]interface{}) {
+			address := addressI.(string)
+			ip := net.ParseIP(address)
+			if ip == nil {
+				return fmt.Errorf("Could not parse addresses '%s'", address)
+			}
+			log.Printf("[INFO] Removing IP/MAC/host=%s/%s/%s in '%s' network", ip.String(), mac, hostname, networkName)
+			if err := removeHost(&network, ip.String(), mac, hostname); err != nil {
+				return err
```
Comment (inercia) [inline libvirt/resource_libvirt_domain.go]: So do we have to `removeHost` every single address? This is a pain...

**Comment by eamonnotoole** [comment]
@inercia RE: removing networks.  I wasn't clear enough in my description.  The problem is when you do the following:

- Create a virtual network and a domain that attaches to that virtual network while specifying dhcp addresses
- Remove the domain but keep the virtual network intact.
- Recreate the domain while specifying the same dhcp addresses as before

In this case currently the domain recreation will fail because the old dhcp entries are still present in the virtual network definition.  Note that we haven't destroyed the network; if we destroy the network at the same time as the domain then there is no issue on recreation.

The first part of the patch removed the dhcp entries from the network when a domain is destroyed to avoid the above error.

The second part of the patch attempts to fix an issue whereby on multiple domain destroy/recreate cycles while retaining the virtual network(s) - after perhaps two of those cycles - we no longer get a complete list of network addresses from domain. ListAllInterfaceAddresses.  In fact, the list is often empty.  In this case the first part of the patch fails to remove the existing dhcp entries (since the address list is empty or incomplete) and domain recreation again fails.  We can still get the list of dhcp entries directly from the virtual network xml, which is what we do

File: libvirt/resource_libvirt_domain.go (hunk 2 above)
Comment (eamonnotoole) [inline libvirt/resource_libvirt_domain.go]: That sounds good, I'll make the change.

Comment (inercia) [inline libvirt/resource_libvirt_domain.go]: Thanks

**Comment by inercia** [comment]
> In this case currently the domain recreation will fail because the old dhcp entries are still present in the virtual network definition. Note that we haven't destroyed the network; if we destroy the network at the same time as the domain then there is no issue on recreation.

Ok @eamonnotoole , thanks for the clarification. In this case, is the virtual network managed by Terraform? I think it is not, right? Because to destroy a domain while leaving the network is not what users usually do, right?

And, anyway, what is the wrong with having the same DHCP mappings in the network when the domain is recreated? It will get the same IP and that is all, right? And that is not a so bad behaviour... Maybe we should try to be smarter when querying the IP address...

Regarding the second part of the PR, we know `ListAllInterfaceAddresses` is not reliable at all. That's the reason why we extended it for using the qemu agent when available (and we recommend to use it). Any contribution for improving the reliability of `getDomainInterfaces` is appreciated...

File: libvirt/resource_libvirt_domain.go (hunk 4 above)
Comment (eamonnotoole) [inline libvirt/resource_libvirt_domain.go]: This section of code is necessary in the case where a domain is destroyed but the network isn't.  I haven't thought of a way of figuring out if that "apply" has been triggered.

File: libvirt/resource_libvirt_domain.go (hunk 1 above)
Comment (eamonnotoole) [inline libvirt/resource_libvirt_domain.go]: Done.

**Comment by eamonnotoole** [comment]
This is something that we do extensively - create virtual networks, and then successively destroy and recreate the domains that are attached to those networks.  We use the same IP addresses each time that we recreate the networks.  The problem is that when you recreate the domain with those same addresses you get a libvirt error that the addresses are already taken.  The only way around this that I've come up with is to remove the dhcp entries from the network definitions (the XML) when you destroy the domain.

Is there an alternative?

It sounds like the method to get interface information from the network definitions is generally useful, which is good to know.

**Comment by inercia** [comment]
> This is something that we do extensively - create virtual networks, and then successively destroy and recreate the domains that are attached to those networks. We use the same IP addresses each time that we recreate the networks. The problem is that when you recreate the domain with those same addresses you get a libvirt error that the addresses are already taken. The only way around this that I've come up with is to remove the dhcp entries from the network definitions (the XML) when you destroy the domain.

So you say that _"when you recreate the domain with those same addresses you get a libvirt error that the addresses are already taken"_. Maybe we should be smarter when we get such an error: in that case we should go further and ask _"ok, and who has that IP?"_ and if the MAC address is the address of the interface I'm working on, then we are fine, right?

> It sounds like the method to get interface information from the network definitions is generally useful, which is good to know.

Yeah, any improvement in the method used for getting addresses is more than welcome.

**Comment by eamonnotoole** [comment]
I've changed the logic for the first part of the patch.  Now before I try to add a domain & IP to the list of dhcp entries for a network I first check to see if the domain is already present in this list, and if so I delete it.  I do things this way since the mac address will change on domain recreation if the mac address hasn't been specified in the .tf, and it seemed easier to just delete the existing entry than to mess around with changing the generated mac back to the the value that I can get from the list of dhcp entries in the network.

**Comment by eamonnotoole** [comment]
@inercia What's the story with this PR?  I've heard nothing in over two months.

**Comment by eamonnotoole** [comment]
@inercia @dmacvicar What's the prognosis with this PR?  I've just rebased, and I think that I've addressed Inercia's issues.

**Comment by MaximilianMeister** [comment]
@inercia I've just built an rpm with this patch, and I can confirm that it fixes the issue. thanks @eamonnotoole

**Comment by kiall** [comment]
When giving machines fixed addresses, this PR somehow causes TF to mix up the IPs:

Two machines, with 10.17.3.0 and 10.17.3.1 as their IPs, but TF always connects to .0.

	libvirt_domain.minion.0 (remote-exec): Connecting to remote host via SSH...
	libvirt_domain.minion.0 (remote-exec):   Host: 10.17.3.0
	libvirt_domain.minion.0 (remote-exec):   User: root
	libvirt_domain.minion.0 (remote-exec):   Password: false
	libvirt_domain.minion.0 (remote-exec):   Private key: true
	libvirt_domain.minion.0 (remote-exec):   SSH Agent: false
	libvirt_domain.minion.1 (remote-exec): Connecting to remote host via SSH...
	libvirt_domain.minion.1 (remote-exec):   Host: 10.17.3.0
	libvirt_domain.minion.1 (remote-exec):   User: root
	libvirt_domain.minion.1 (remote-exec):   Password: false
	libvirt_domain.minion.1 (remote-exec):   Private key: true
	libvirt_domain.minion.1 (remote-exec):   SSH Agent: false

**Comment by eamonnotoole** [comment]
Kiall, I'll have a look at those IP issues next week.

**Comment by eamonnotoole** [comment]
@inercia @kiall @MaximilianMeister I've fixed that bug where the TF doesn't use the correct address for multiple domains.  This is what I see now:

```console
libvirt_domain.domain.1 (remote-exec): Connecting to remote host via SSH...
libvirt_domain.domain.1 (remote-exec):   Host: 192.168.0.3
libvirt_domain.domain.1 (remote-exec):   User: root
libvirt_domain.domain.1 (remote-exec):   Password: false
libvirt_domain.domain.1 (remote-exec):   Private key: false
libvirt_domain.domain.1 (remote-exec):   SSH Agent: true
libvirt_domain.domain.0 (remote-exec): Connecting to remote host via SSH...
libvirt_domain.domain.0 (remote-exec):   Host: 192.168.0.2
libvirt_domain.domain.0 (remote-exec):   User: root
libvirt_domain.domain.0 (remote-exec):   Password: false
libvirt_domain.domain.0 (remote-exec):   Private key: false
libvirt_domain.domain.0 (remote-exec):   SSH Agent: true
libvirt_domain.domain.1 (remote-exec): Connecting to remote host via SSH...
libvirt_domain.domain.1 (remote-exec):   Host: 192.168.0.3
libvirt_domain.domain.1 (remote-exec):   User: root
libvirt_domain.domain.1 (remote-exec):   Password: false
libvirt_domain.domain.1 (remote-exec):   Private key: false
libvirt_domain.domain.1 (remote-exec):   SSH Agent: true
libvirt_domain.domain.0 (remote-exec): Connecting to remote host via SSH...
libvirt_domain.domain.0 (remote-exec):   Host: 192.168.0.2
libvirt_domain.domain.0 (remote-exec):   User: root
libvirt_domain.domain.0 (remote-exec):   Password: false
libvirt_domain.domain.0 (remote-exec):   Private key: false
libvirt_domain.domain.0 (remote-exec):   SSH Agent: true
libvirt_domain.domain.1 (remote-exec): Connecting to remote host via SSH...
libvirt_domain.domain.1 (remote-exec):   Host: 192.168.0.3
libvirt_domain.domain.1 (remote-exec):   User: root
libvirt_domain.domain.1 (remote-exec):   Password: false
libvirt_domain.domain.1 (remote-exec):   Private key: false
libvirt_domain.domain.1 (remote-exec):   SSH Agent: true
libvirt_domain.domain.0 (remote-exec): Connecting to remote host via SSH...
libvirt_domain.domain.0 (remote-exec):   Host: 192.168.0.2
libvirt_domain.domain.0 (remote-exec):   User: root
libvirt_domain.domain.0 (remote-exec):   Password: false
libvirt_domain.domain.0 (remote-exec):   Private key: false
libvirt_domain.domain.0 (remote-exec):   SSH Agent: true
```

**Comment by kiall** [comment]
<aside>It's only seeing the GitHub notification emails a few minutes ago that I made the connection you were the same @eamonnotoole from HPE :)</aside>

I'll happily test this out and verify the issue I found is gone, but can't do much more myself (I know just about zero re this code ;))

**Comment by eamonnotoole** [comment]
@kiall Thanks.  Small world.

**Comment by dmacvicar** [comment]
@kiall did you had the chance to test this?

File: libvirt/resource_libvirt_domain_test.go [hunk 5]
```diff
@@ -540,3 +599,73 @@ func testAccCheckLibvirtScsiDisk(n string, domain *libvirt.Domain) resource.Test
...
+		networkDef, err := newDefNetworkfromLibvirt(retrieveNetwork)
+
+		for _, ips := range networkDef.Ips {
+			for _, dhcpHost := range ips.Dhcp.Hosts {
+				if dhcpHost.Ip == ip {
+					return nil
+				}
+			}
+		}
+		return fmt.Errorf("Hostname with ip '%s' does not have a dhcp entry in network", ip)
```
Comment (MalloZup) [inline libvirt/resource_libvirt_domain_test.go]: here, we 2 returns, the 2nd one will never called

File: libvirt/resource_libvirt_domain.go [hunk 6]
```diff
@@ -1061,7 +1073,8 @@ func newDiskForCloudInit(virConn *libvirt.Connect, volumeKey string) (defDisk, e
 	return disk, nil
 }
 
-func getDomainInterfaces(domain libvirt.Domain) ([]libvirt.DomainInterface, error) {
+func getDomainInterfaces(domain libvirt.Domain, domainDef defDomain,
```
Comment (MalloZup) [inline libvirt/resource_libvirt_domain.go]: what is `defDomain`?

**Comment by MalloZup** [comment]
afaik we cannot anymore merge this patch as it is.

Resolving the conflicts, we get then afterwards:

```console
go build
# github.com/dmacvicar/terraform-provider-libvirt/libvirt
libvirt/resource_libvirt_domain.go:1127:12: undefined: defDomain
libvirt/resource_libvirt_domain.go:1163:30: undefined: defNetworkInterface
libvirt/resource_libvirt_domain.go:1164:35: undefined: defDomain
libvirt/resource_libvirt_domain.go:1266:59: undefined: defDomain
libvirt/resource_libvirt_domain.go:1305:45: undefined: defDomain
libvirt/resource_libvirt_domain.go:1327:33: networkDef.Ips undefined (type libvirtxml.Network has no field or method Ips, but does have IPs)
libvirt/utils_libvirt.go:49:52: undefined: defNetwork
Makefile:4: recipe for target 'build' failed
```
From my pov, maybe this defNetworks defDomain where existing at the time

**Comment by eamonnotoole** [comment]
@dmacvicar @MalloZup I've rebased the patch, and removed that redundant return.

**Comment by MalloZup** [comment]
looks good to me

**Comment by MalloZup** [comment]
@inercia feel free to say the lastword review to the rebased one pr ,since you @ begin

**Comment by eamonnotoole** [comment]
@MalloZup @inercia @dmacvicar I rebased, it turns out that a recently merged change a20d2a9 negates the need for removing hosts, so that part of the patch is gone.  The patch now implements an alternative way of getting information on domain interfaces.

**Comment by dmacvicar** [comment]
@MaximilianMeister @inercia is this ok to be merged now?

**Pull request merged**